File: backend/controllers/contactosController.py
from flask import request, jsonify
from models.contactosModel import ContactosModel
from datetime import datetime


def get_all_contactos():
    try:
        contactos = ContactosModel.get_all()

        # Convertir datetime a string para JSON
        for contacto in contactos:
            if contacto.get('created_at'):
                contacto['created_at'] = contacto['created_at'].isoformat()
            if contacto.get('updated_at'):
                contacto['updated_at'] = contacto['updated_at'].isoformat()
            if contacto.get('deleted_at'):
                contacto['deleted_at'] = contacto['deleted_at'].isoformat()

        return jsonify({
            'success': True,
            'data': contactos,
            'message': 'Contactos obtenidos exitosamente'
        }), 200

    except Exception as e:
        return jsonify({
            'success': False,
            'message': f'Error al obtener contactos: {str(e)}'
        }), 500


def get_contacto_by_id(contacto_id):
    try:
        contacto = ContactosModel.get_by_id(contacto_id)

        if not contacto:
            return jsonify({
                'success': False,
                'message': 'Contacto no encontrado'
            }), 404

        # Convertir datetime a string para JSON
        if contacto.get('created_at'):
            contacto['created_at'] = contacto['created_at'].isoformat()
        if contacto.get('updated_at'):
            contacto['updated_at'] = contacto['updated_at'].isoformat()
        if contacto.get('deleted_at'):
            contacto['deleted_at'] = contacto['deleted_at'].isoformat()

        return jsonify({
            'success': True,
            'data': contacto,
            'message': 'Contacto obtenido exitosamente'
        }), 200

    except Exception as e:
        return jsonify({
            'success': False,
            'message': f'Error al obtener el contacto: {str(e)}'
        }), 500


def create_contacto():
    try:
        data = request.get_json()

        # Validaciones básicas
        if not data:
            return jsonify({
                'success': False,
                'message': 'No se enviaron datos'
            }), 400

        # Verificar si ya existe un contacto con el mismo correo
        # if data.get('correo'):
        #     existing_contacto = ContactosModel.get_by_email(data['correo'])
        #     if existing_contacto:
        #         return jsonify({
        #             'success': False,
        #             'message': 'Ya existe un contacto con este correo electrónico'
        #         }), 409

        # Crear nuevo contacto
        contacto_id = ContactosModel.create(data)

        if contacto_id:
            # Obtener el contacto creado para devolverlo
            nuevo_contacto = ContactosModel.get_by_id(contacto_id)

            # Convertir datetime a string para JSON
            if nuevo_contacto.get('created_at'):
                nuevo_contacto['created_at'] = nuevo_contacto['created_at'].isoformat(
                )
            if nuevo_contacto.get('updated_at'):
                nuevo_contacto['updated_at'] = nuevo_contacto['updated_at'].isoformat(
                )

            return jsonify({
                'success': True,
                'data': nuevo_contacto,
                'message': 'Contacto creado exitosamente'
            }), 201
        else:
            return jsonify({
                'success': False,
                'message': 'Error al crear el contacto'
            }), 500

    except Exception as e:
        return jsonify({
            'success': False,
            'message': f'Error al crear contacto: {str(e)}'
        }), 500


def update_contacto(contacto_id):
    try:
        # Verificar si el contacto existe
        contacto_existente = ContactosModel.get_by_id(contacto_id)
        if not contacto_existente:
            return jsonify({
                'success': False,
                'message': 'Contacto no encontrado'
            }), 404

        data = request.get_json()
        if not data:
            return jsonify({
                'success': False,
                'message': 'No se enviaron datos para actualizar'
            }), 400

        # Verificar si el correo ya existe en otro contacto
        # if data.get('correo') and data['correo'] != contacto_existente.get('correo'):
        #     existing_contacto = ContactosModel.get_by_email(data['correo'])
        #     if existing_contacto and existing_contacto['idContacto'] != contacto_id:
        #         return jsonify({
        #             'success': False,
        #             'message': 'Ya existe otro contacto con este correo electrónico'
        #         }), 409

        # Actualizar contacto
        updated = ContactosModel.update(contacto_id, data)

        if updated:
            # Obtener el contacto actualizado
            contacto_actualizado = ContactosModel.get_by_id(contacto_id)

            # Convertir datetime a string para JSON
            if contacto_actualizado.get('created_at'):
                contacto_actualizado['created_at'] = contacto_actualizado['created_at'].isoformat(
                )
            if contacto_actualizado.get('updated_at'):
                contacto_actualizado['updated_at'] = contacto_actualizado['updated_at'].isoformat(
                )

            return jsonify({
                'success': True,
                'data': contacto_actualizado,
                'message': 'Contacto actualizado exitosamente'
            }), 200
        else:
            return jsonify({
                'success': False,
                'message': 'Error al actualizar el contacto'
            }), 500

    except Exception as e:
        return jsonify({
            'success': False,
            'message': f'Error al actualizar contacto: {str(e)}'
        }), 500


def delete_contacto(contacto_id):
    try:
        # Verificar si el contacto existe
        contacto_existente = ContactosModel.get_by_id(contacto_id)
        if not contacto_existente:
            return jsonify({
                'success': False,
                'message': 'Contacto no encontrado'
            }), 404

        # Eliminar contacto
        deleted = ContactosModel.delete(contacto_id)

        if deleted:
            return jsonify({
                'success': True,
                'message': 'Contacto eliminado exitosamente'
            }), 200
        else:
            return jsonify({
                'success': False,
                'message': 'Error al eliminar el contacto'
            }), 500

    except Exception as e:
        return jsonify({
            'success': False,
            'message': f'Error al eliminar contacto: {str(e)}'
        }), 500

File: backend/controllers/contactosController.js
const contactosModel = require('../models/contactosModel');

// 🧠 Validaciones específicas de Perú
const validarDatos = (data) => {
  const errores = [];

  // RUC: debe tener 11 dígitos numéricos
  if (!/^\d{11}$/.test(data.rucEmpresa)) {
    errores.push('El RUC debe tener 11 dígitos numéricos');
  }

  // Nombre y apellido obligatorios
  if (!data.nombre || data.nombre.trim() === '') {
    errores.push('El nombre es obligatorio');
  }

  if (!data.apellido || data.apellido.trim() === '') {
    errores.push('El apellido es obligatorio');
  }

  // Teléfono y WhatsApp: 9 dígitos y empezar con 9
  const telefonoValido = /^\d{9}$/.test(data.telefono) && data.telefono.startsWith('9');
  const whatsappValido = /^\d{9}$/.test(data.whatsapp) && data.whatsapp.startsWith('9');

  if (!telefonoValido) {
    errores.push('El teléfono debe tener 9 dígitos y empezar con 9');
  }

  if (!whatsappValido) {
    errores.push('El WhatsApp debe tener 9 dígitos y empezar con 9');
  }

  // Correo (opcional pero si está, debe ser válido)
  if (data.correo && !/^[^\s@]+@[^\s@]+\.[^\s@]+$/.test(data.correo)) {
    errores.push('El correo no es válido');
  }

  return errores;
};

const getAllContactos = async (req, res) => {
  try {
    const contactos = await contactosModel.getAll();
    res.status(200).json(contactos);
  } catch (error) {
    console.error('Error al obtener los contactos:', error);
    res.status(500).json({ error: 'Error al obtener los contactos' });
  }
};

const getContactoById = async (req, res) => {
  const { id } = req.params;
  try {
    const contacto = await contactosModel.getById(id);
    if (!contacto) {
      return res.status(404).json({ error: 'Contacto no encontrado' });
    }
    res.status(200).json(contacto);
  } catch (error) {
    console.error('Error al obtener el contacto:', error);
    res.status(500).json({ error: 'Error al obtener el contacto' });
  }
};

const createContacto = async (req, res) => {
  const data = req.body;
  const errores = validarDatos(data);
  if (errores.length > 0) {
    return res.status(400).json({ errores });
  }

  try {
    const newContactoId = await contactosModel.create(data);
    res.status(201).json({ id: newContactoId, message: 'Contacto creado exitosamente' });
  } catch (error) {
    console.error('Error al crear el contacto:', error);
    res.status(500).json({ error: 'Error al crear el contacto' });
  }
};

const updateContacto = async (req, res) => {
  const { id } = req.params;
  const data = req.body;
  const errores = validarDatos(data);
  if (errores.length > 0) {
    return res.status(400).json({ errores });
  }

  try {
    const actualizado = await contactosModel.update(id, data);
    if (!actualizado) {
      return res.status(404).json({ error: 'Contacto no encontrado' });
    }
    res.status(200).json({ message: 'Contacto actualizado correctamente' });
  } catch (error) {
    console.error('Error al actualizar el contacto:', error);
    res.status(500).json({ error: 'Error al actualizar el contacto' });
  }
};

const deleteContacto = async (req, res) => {
  const { id } = req.params;
  try {
    const eliminado = await contactosModel.remove(id);
    if (!eliminado) {
      return res.status(404).json({ error: 'Contacto no encontrado' });
    }
    res.status(200).json({ message: 'Contacto eliminado correctamente' });
  } catch (error) {
    console.error('Error al eliminar el contacto:', error);
    res.status(500).json({ error: 'Error al eliminar el contacto' });
  }
};

module.exports = {
  getAllContactos,
  getContactoById,
  createContacto,
  updateContacto,
  deleteContacto
};